source/linker.cpp

```diff
@@ -0,0 +1,1098 @@
+  case spv::Op::OpTypeEvent:
+  case spv::Op::OpTypeDeviceEvent:
+  case spv::Op::OpTypeReserveId:
+  case spv::Op::OpTypeQueue:
+  case spv::Op::OpTypePipe:
+  case spv::Op::OpTypeForwardPointer:
+  case spv::Op::OpConstantTrue:
+  case spv::Op::OpConstantFalse:
+  case spv::Op::OpConstant:
+  case spv::Op::OpConstantComposite:
+  case spv::Op::OpConstantSampler:
+  case spv::Op::OpConstantNull:
+  case spv::Op::OpSpecConstantTrue:
+  case spv::Op::OpSpecConstantFalse:
+  case spv::Op::OpSpecConstant:
+  case spv::Op::OpSpecConstantComposite:
+  case spv::Op::OpSpecConstantOp:
+  case spv::Op::OpVariable:
+  case spv::Op::OpUndef:
+    return true;
+  default:
+    return false;
+  }
+}
+
+static spv_result_t MergeModules(spv_context_t& context,
+                                 std::vector<std::vector<uint32_t>> &binaries,
+                                 size_t num_binaries, uint32_t max_index,
+                                 uint32_t version,
+                                 spv_binary_t& merged_spvbinary,
+                                 bool create_library) {
+  spv_position_t position = {};
+
+  std::vector<uint32_t> indices(num_binaries, 5u);
+
+  std::unordered_map<spv::Id, const char *> exports;
+  std::unordered_map<spv::Id, const char *> imports;
+  std::unordered_map<spv::Id, spv::Id> link_table;
+  std::vector<spv::Id> builtins;
+
+  uint32_t* merged_binary = merged_spvbinary.code;
+  size_t& merged_length = merged_spvbinary.wordCount;
+
+  // Generate new header
+  SetWord(merged_binary, 0u, spv::MagicNumber);
+  SetWord(merged_binary, 1u, version);
+  SetWord(merged_binary, 2u, 0u);
+  SetWord(merged_binary, 3u, max_index);
+  SetWord(merged_binary, 4u, 0u);
+  merged_length = 5u;
+
+  const auto CopySections = [&binaries,&indices,num_binaries,
+                             &merged_binary,&merged_length](const std::function<bool (spv::Op)>& check_section) {
+    for (uint32_t i = 0u; i < num_binaries; ++i) {
+      const uint32_t* const binary = binaries[i].data();
+      const size_t length = binaries[i].size();
+      const uint32_t index = indices[i];
+      uint32_t j = index;
+      uint32_t prev_j = j;
+
+      while (j < length) {
+        uint32_t word_count;
+        const spv::Op opcode = ParseOpcode(binary, j, word_count);
+
+        prev_j = j;
+
+        if (!check_section(opcode))
+          break;
+
+        j += word_count;
+      }
+      if (j == length)
+        prev_j = j;
+
+      std::memcpy(merged_binary + merged_length, binary + index,
+                  (prev_j - index) * sizeof(spirv_word));
+      merged_length += prev_j - index;
+      indices[i] = j;
+    }
+  };
+
+  // Merge all OpCapability
+  std::unordered_set<uint32_t> capabilities;
+  for (uint32_t i = 0u; i < num_binaries; ++i) {
+    const uint32_t* const binary = binaries[i].data();
+    const size_t length = binaries[i].size();
+    const uint32_t index = indices[i];
+    uint32_t j = index;
+    uint32_t prev_j = j;
```

Comment:
That is very true… a leftover from copy/pasting that loop structure from somewhere else. I am a bit surprised that clang did not pick it up locally. Similarly to all the errors/warning thrown by Travis. I’ll add locally some more warning flags to the default ones./